state-chain/pallets/cf-broadcast/src/migrations/v2.rs

```diff
@@ -2,47 +2,168 @@ use crate::*;
+			if let Some(attempt) = old::AwaitingBroadcast::<T, I>::get(old::BroadcastAttemptId {
+				broadcast_id,
+				attempt_count,
+			}) {
+				AwaitingBroadcast::<T, I>::insert(
+					broadcast_id,
+					TransactionSigningAttempt {
+						broadcast_attempt: BroadcastAttempt {
+							broadcast_id,
+							transaction_payload: attempt.broadcast_attempt.transaction_payload,
+							threshold_signature_payload: attempt
+								.broadcast_attempt
+								.threshold_signature_payload,
+							transaction_out_id: attempt.broadcast_attempt.transaction_out_id,
+						},
+						nominee: attempt.nominee,
+					},
+				);
+			}
+		});
+
+		// Failed broadcaster storage: Vec<ValidatorId> -> BTreeSet<ValidatorId>
+		old::FailedBroadcasters::<T, I>::drain().for_each(|(broadcast_id, failed_broadcasters)| {
+			FailedBroadcasters::<T, I>::insert(
+				broadcast_id,
+				BTreeSet::from_iter(failed_broadcasters),
+			)
+		});
+
+		let retry_queue = old::BroadcastRetryQueue::<T, I>::take()
+			.unwrap_or_default()
+			.into_iter()
+			.map(|attempt| BroadcastAttempt::<T, I> {
+				broadcast_id: attempt.broadcast_attempt_id.broadcast_id,
+				transaction_payload: attempt.transaction_payload,
+				threshold_signature_payload: attempt.threshold_signature_payload,
+				transaction_out_id: attempt.transaction_out_id,
+			})
+			.collect::<Vec<_>>();
+		BroadcastRetryQueue::<T, I>::put(retry_queue);
+
 		Weight::zero()
 	}
 
 	#[cfg(feature = "try-runtime")]
 	fn pre_upgrade() -> Result<Vec<u8>, DispatchError> {
-		Ok(Default::default())
+		let broadcasts_with_failed_broadcasters =
+			old::FailedBroadcasters::<T, I>::iter_keys().collect::<Vec<_>>();
+		let broadcast_retry_queue = old::BroadcastRetryQueue::<T, I>::get()
+			.unwrap_or_default()
+			.into_iter()
+			.map(|attempt| attempt.broadcast_attempt_id.broadcast_id)
+			.collect::<Vec<_>>();
+		Ok(MigrationVerification { broadcasts_with_failed_broadcasters, broadcast_retry_queue }
+			.encode())
 	}
 
 	#[cfg(feature = "try-runtime")]
-	fn post_upgrade(_state: Vec<u8>) -> Result<(), DispatchError> {
+	fn post_upgrade(state: Vec<u8>) -> Result<(), DispatchError> {
 		let pending_broadcasts = PendingBroadcasts::<T, I>::get();
 		for id in AwaitingBroadcast::<T, I>::iter_keys() {
-			assert!(pending_broadcasts.contains(&id.broadcast_id),);
+			assert!(pending_broadcasts.contains(&id));
 		}
+
+		let MigrationVerification { broadcasts_with_failed_broadcasters, broadcast_retry_queue } =
+			<MigrationVerification>::decode(&mut &state[..]).unwrap();
+
+		broadcasts_with_failed_broadcasters.into_iter().for_each(|broadcast_id| {
+			assert!(FailedBroadcasters::<T, I>::contains_key(broadcast_id))
+		});
+
+		let new_retry_queue = BroadcastRetryQueue::<T, I>::get()
+			.into_iter()
+			.map(|attempt| attempt.broadcast_id)
+			.collect::<Vec<_>>();
+
+		broadcast_retry_queue
+			.into_iter()
+			.for_each(|broadcast_id| assert!(new_retry_queue.contains(&broadcast_id)));
+
```

Comment:
we can add some unit tests for these migrations too - a good habit to get into

Reply:
Added unit tests for migration. I didn't even know we could unit test this but having these test makes me much more confidant about the migration!